REGRESSION(258522@main): Ensure the slider style is applied to an <input> element before drawing

shallawa opened the pull request:
#### c14ea6ce8716cee385610874385c04417621068c
<pre>
REGRESSION(258522@main): Ensure the slider style is applied to an &lt;input&gt; element before drawing
<a href="https://bugs.webkit.org/show_bug.cgi?id=250334">https://bugs.webkit.org/show_bug.cgi?id=250334</a>
rdar://104019874

Reviewed by Aditya Keerthi.

The appearance style can be applied to any non &lt;input&gt; element. In 258522@main,
createSliderTrackPartForRenderer() was refactored from RenderTheme::paintSliderTicks()
which makes an early return &quot;if (!is&lt;HTMLInputElement&gt;(o.node()))&quot;. We need to
port this check to createSliderTrackPartForRenderer().

* LayoutTests/fast/forms/slider-track-not-input-element-expected.txt: Added.
* LayoutTests/fast/forms/slider-track-not-input-element.html: Added.
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::createSliderTrackPartForRenderer):

Canonical link: <a href="https://commits.webkit.org/258684@main">https://commits.webkit.org/258684@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/26b131af2487845ebe99c349797d8314a90d119e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102673 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11798 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/35704 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111936 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/172181 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106640 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12806 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2704 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/94940 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/109632 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108449 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/9827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/35704 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/94940 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91679 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/35704 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/94940 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5255 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/35704 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/5411 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/2418 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/11432 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/35704 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7145 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on LayoutTests/fast/forms/slider-track-not-input-element.html:
```diff
@@ -0,0 +1,12 @@
+	<p>This test passes if it does not crash.</p>
+	<script>
+		if (window.testRunner)
+    		testRunner.dumpAsText();
```
Nit: Indentation.

shallawa replied:
Tabs were replaced by spaces.

pxlcoder commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -503,6 +503,9 @@ static RefPtr<ControlPart> createProgressBarPartForRenderer(const RenderObject&
 
 static RefPtr<ControlPart> createSliderTrackPartForRenderer(const RenderObject& renderer)
 {
+    if (!is<HTMLInputElement>(renderer.node()))
```
You can use `dynamicDowncast` in line 513 to better organize the logic.

shallawa replied:
Done.

pxlcoder reviewed (commented):
A longer term solution would be to resolve to `appearance: none` here, if the element is not an `<input type=range>` but this approach is fine for now, since it restores the existing behavior.

shallawa committed:
REGRESSION(258522@main): Ensure the slider style is applied to an <input> element before drawing
https://bugs.webkit.org/show_bug.cgi?id=250334
rdar://104019874

Reviewed by Aditya Keerthi.

The appearance style can be applied to any non <input> element. In 258522@main,
createSliderTrackPartForRenderer() was refactored from RenderTheme::paintSliderTicks()
which makes an early return "if (!is<HTMLInputElement>(o.node()))". We need to
port this check to createSliderTrackPartForRenderer().

* LayoutTests/fast/forms/slider-track-not-input-element-expected.txt: Added.
* LayoutTests/fast/forms/slider-track-not-input-element.html: Added.
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::createSliderTrackPartForRenderer):

Canonical link: https://commits.webkit.org/258684@main
